M211 command

d00616 opened the pull request:
This command is to disable software endstops per axis. http://reprap.org/wiki/G-code#M211:_Disable.2FEnable_software_endstops

d00616 committed:
M211 command

thinkyhead commented on Marlin/Configuration_adv.h:
```diff
@@ -90,6 +90,10 @@
+// Force this when M211 command is enabled
+#if ENABLE_M211 == true
+  #define ENDSTOPS_ONLY_FOR_HOMING
+#endif
```
Things like this should go into `Conditionals.h` now.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5227,22 +5290,36 @@ void get_arc_coordinates() {
-    if (target[X_AXIS] < min_pos[X_AXIS]) target[X_AXIS] = min_pos[X_AXIS];
-    if (target[Y_AXIS] < min_pos[Y_AXIS]) target[Y_AXIS] = min_pos[Y_AXIS];
-    
+    #if ((!defined(ENABLE_M211)) || (ENABLE_M211 == false))
```
Should just be `#ifndef ENABLE_M211`.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -369,6 +369,10 @@ const bool Z_PROBE_ENDSTOP_INVERTING = false; // set to true to invert the logic
 
+// Allow M211 command to disable software endstops temporarly. This enabled ENDSTOPS_ONLY_FOR_HOMING
+// In combination with G92 command marlin can be used to manually do an X,Y,Z axis homing and mill down an part
+#define ENABLE_M211 true
```
Should just be `#define ENABLE_M211`.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -292,7 +292,15 @@
   /**
+   * Check if ENDSTOPS_ONLY_FOR_HOMING is disabled when M211 command is enabled
+   */
+  #if (ENABLE_M211 == true) && (!defined(ENDSTOPS_ONLY_FOR_HOMING))
```
`#if defined(ENABLE_M211) && !defined(ENDSTOPS_ONLY_FOR_HOMING)` is fine here.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -5227,22 +5290,36 @@ void get_arc_coordinates() {
     #endif
     
-    if (target[Z_AXIS] < min_pos[Z_AXIS]+negative_z_offset) target[Z_AXIS] = min_pos[Z_AXIS]+negative_z_offset;
+    #if ((!defined(ENABLE_M211)) || (ENABLE_M211 == false))
```
`#ifndef …`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4995,6 +5052,12 @@ void process_commands() {
           break;
       #endif
 
+      #if ENABLE_M211 == true
```
`#ifdef …`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3767,6 +3776,54 @@ inline void gcode_M206() {
+/**
+ * M211: Enable or disable software endstops
+ */
+#if ENABLE_M211 == true
```
`#ifdef …`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -216,6 +217,10 @@ float home_offset[3] = { 0 };
 float min_pos[3] = { X_MIN_POS, Y_MIN_POS, Z_MIN_POS };
 float max_pos[3] = { X_MAX_POS, Y_MAX_POS, Z_MAX_POS };
 bool axis_known_position[3] = { false };
+#if ENABLE_M211 == true
```
`#ifdef …`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -151,6 +151,7 @@
 // M207 - Set retract length S[positive mm] F[feedrate mm/min] Z[additional zlift/hop], stays in mm regardless of M200 setting
 // M208 - Set recover=unretract length S[positive mm surplus to the M207 S*] F[feedrate mm/sec]
 // M209 - S<1=true/0=false> enable automatic retract detect if the slicer did not support G10/11: every normal extrude-only move will be classified as retract depending on the direction.
+// M211 - Disable software endstops. This allows using negative coordinates for milling.
```
You can already use negative coordinates. You should make it clear this means something else, like being allowed to go below `Z_MIN_POS`, even if it is set to -20.0mm.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3767,6 +3776,54 @@ inline void gcode_M206() {
+ * M211: Enable or disable software endstops
+ */
+#if ENABLE_M211 == true
+inline void gcode_M211() {
```
You can indent the whole function between the `#ifdef` and `#endif`. The complete coding standards are written up (finally) at https://github.com/MarlinFirmware/Marlin/wiki/Coding-Standards

d00616 commented:
@thinkyhead I have changed my code. Thanks for your comments.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3767,6 +3776,54 @@ inline void gcode_M206() {
+  if (code_seen_str("Z0")) smask |= 4;
+  if (code_seen_str("X1")) smask |= 16;
+  if (code_seen_str("Y1")) smask |= 32;
+  if (code_seen_str("Z1")) smask |= 64;
```
Normally this would be something along the lines of:

```
if (code_seen('X')) smask |= code_value_long() ? 0x10 : 0x01;
if (code_seen('Y')) smask |= code_value_long() ? 0x20 : 0x02;
if (code_seen('Z')) smask |= code_value_long() ? 0x40 : 0x04;
```

...which should run faster, though it might take more SRAM. I would compare. But, your code will allow `M211 X0 X1`. Is that a legal input?

d00616 replied:
Setting multiple directions at same time is my understanding of command description. But it's looks like there is no implemtation of this so i think it possible to limit this command to one direction per axis. http://reprap.org/wiki/G-code#M211:_Disable.2FEnable_software_endstops

thinkyhead replied:
With the latest macros, this might be better:

```
if (code_seen('X')) smask |= code_value_long() ? BIT(4+X_AXIS) : BIT(X_AXIS);
if (code_seen('Y')) smask |= code_value_long() ? BIT(4+Y_AXIS) : BIT(Y_AXIS);
if (code_seen('Z')) smask |= code_value_long() ? BIT(4+Z_AXIS) : BIT(Z_AXIS);
```

Perhaps even this:

```
for (uint8_t i=X_AXIS; i<=Z_AXIS; i++)
  if (code_seen(axis_codes[i])) smask |= code_value_long() ? BIT(i+4) : BIT(i);
```

d00616 replied:
Both changes may be better but it's not possible to allow "M211 X0 X1". See discussion above.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -3767,6 +3776,54 @@ inline void gcode_M206() {
+  SERIAL_PROTOCOLPGM(MSG_Z_MIN);
+  SERIAL_PROTOCOLLN(((enabled_software_endstops & 4) > 0?MSG_ENABLED:MSG_DISABLED));
+  SERIAL_PROTOCOLPGM(MSG_Z_MAX);
+  SERIAL_PROTOCOLLN(((enabled_software_endstops & 64) > 0?MSG_ENABLED:MSG_DISABLED));
```
There should probably be an enum for these bit masks to get rid of "magic numbers."

```
enum EndStopBit { ESMinBitX = 0x01, ESMinBitY = 0x02, ESMinBitZ = 0x04, ESMaxBitX = 0x10, ESMaxBitY = 0x20, ESMaxBitZ = 0x40, ESAllBits = 0x77 };
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -292,7 +292,15 @@
+   * Check if ENDSTOPS_ONLY_FOR_HOMING is disabled when M211 command is enabled
+   */
+  #if (defined(ENABLE_M211)) && (!defined(ENDSTOPS_ONLY_FOR_HOMING))
+    #error ENDSTOPS_ONLY_FOR_HOMING must be enabled when M211 command is active.
```
Good to have, but since `ENDSTOPS_ONLY_FOR_HOMING` is enabled in `Conditionals.h` when `ENABLE_M211` is enabled, this error will never be thrown.

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -292,7 +292,15 @@
+
+  /**
    * Warnings for old configurations
+   * Check if ENDSTOPS_ONLY_FOR_HOMING is disabled when M211 command is enabled
```
Oops, this line is a repeat of the one above.

nophead replied:
Why is it necessary to disable the soft end stops when you move the origin
to match the work-piece? They should still apply to the machine coordinates
after they are offset from the model coordinates as they are the physical
limits of the machine, which don't change, just as physical ones don't.

For example, my CNC mill running LinuxCNC has endstop switches at one end
of each axis and a soft limit to stop it crashing at the other end. They
always apply but I can touch off the work-piece and change the origin to
anywhere. The limits still apply so I cant crash an axis although I can of
course crash the tool.

Possibly Marlin applies the soft end stops at the wrong place but I think
this is the wrong way to fix it.

d00616 replied:
I want to use marlin for milling and printing with only one configuration. In printing mode i can home the Z axis automatically with an servo. In milling mode i need manual home the Z axis.

To do manual homing and milling down i need to allow moving Z under 0 position. I can allow this all the time or, with this command only when i need this.

I want to block moves under 0 when an extruder is mounted to protect my hardware when i send an milling file to marlin.

My first idea was to implement secondary homing (G30) but this command is defined with another functionality in marlin and needs to add an secondary pointer set for X,Y and Z.

tonnico replied:
Or you can just set Z to 10 by G92 Z10. If it's at zero you can now go down by 10mm.

d00616 replied:
> In Marlin/SanityCheck.h
> https://github.com/MarlinFirmware/Marlin/pull/1814#discussion_r28127276:
> 
> > ```
> > * Warnings for old configurations
> > ```
> > -   \* Check if ENDSTOPS_ONLY_FOR_HOMING is disabled when M211 command is enabled
> 
> Or you can just set Z to 10 by G92 Z10. If it's at zero you can now go
> down by 10mm.
> 
> That's an option but needs g-code dealing between 0mm and 10mm in
> difference to g-code working between -10mm and 0mm.

This command fulfil my requirements. If there is a wish to integrate
this command in marlin i clean up the code. Other way i let this code in
my fork of marlin.

boelle commented:
Please rebase, the code cant be merged as is

thinkyhead committed:
Merge pull request #2269 from Wackerbarth/SwitchToArduinoIDE1.6.4

Switch to Arduino IDE 1.6.4  for Travis testing

thinkyhead committed:
Cleanup of comments & spacing

thinkyhead committed:
Fix comment on M83, add comment to prepare_move

Wackerbarth committed:
Clean up comments and spacing

AnHardt committed:
Reactivate tests for more extruders in Travis

BOARD_ULTIMAKER was not the default any more
- sed failed to change the board
- RAMPS_EFB has no second extruder
- ...
- Added test for bed

Rebased.

Wackerbarth committed:
Merge pull request #2348 from AnHardt/n-Travis-MoreExtruders

Reactivate tests for more extruders in Travis

AnHardt committed:
Add MagoKimbra to the list of contributors

in Development

boelle committed:
Merge pull request #2356 from AnHardt/MK

Add MagoKimbra to the list of contributors

d00616 commented:
rebased and squashed but untested. I plan to test it in a frew days when i ported my configuration to actual branch.

d00616 commented:
Tested successfully. In my opinion this code can be merged.

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -387,7 +387,9 @@ const bool Z_PROBE_ENDSTOP_INVERTING = false; // set to true to invert the logic
 #define min_software_endstops true // If true, axis won't move to coordinates less than HOME_POS.
 #define max_software_endstops true  // If true, axis won't move to coordinates greater than the defined lengths below.
 
-// @section machine
```
Probably good to keep this section directive, even though the configurator is currently stale.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -4220,6 +4229,54 @@ inline void gcode_M206() {
+    // Print state
+  SERIAL_PROTOCOLLN(MSG_M211_REPORT);
+  SERIAL_PROTOCOLPGM(MSG_X_MIN);
+  SERIAL_PROTOCOLLN(((enabled_software_endstops & 1) > 0?MSG_ENABLED:MSG_DISABLED));
```
We also have a handy `TEST` macro that can make these slightly clearer, e.g.:

```
SERIAL_PROTOCOLLN(TEST(enabled_software_endstops, 0) ? MSG_ENABLED : MSG_DISABLED);
```

d00616 committed:
PR notes (@thinkyhead)

d00616 commented:
@Wackerbarth Thank you for closing this PR without comments. I my opinion this is not respectfulness about many hours I spend into design, implementation and rebasing this work. I don't share my work in future with this project.

Don't reopen this PR. I'm not interested wasting more of my time by contributing to Marlin.

Wackerbarth commented:
@d00616 -- I don't know how "I" closed this. It is a very old PR and I don't think that I have ever looked at it. This is the second PR that has mysteriously appeared as closed. In any case, it is rather outdated and needs updating to the current "master" branch on MarlinDev.
